### PR TITLE
Adds warning if tests don't provide valid variable

### PR DESCRIPTION
**Pull request by Andrew-Hayes**
Resolves https://github.com/opentofu/opentofu/issues/2012

Adds a warning to `tofu test` if we find variable validation errors in the plan stage of a test that are expected in the apply stage. 

## Target Release

1.9.0

## Checklist

- [x] I have read the [contribution guide](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md).
- [x] I have not used an AI coding assistant to create this PR.
- [x] I have written all code in this PR myself OR I have marked all code I have not written myself (including modified code, e.g. copied from other places and then modified) with a comment indicating where it came from.
- [x] I (and other contributors to this PR) have not looked at the Terraform source code while implementing this PR.

### Go checklist

 

- [x] I have run golangci-lint on my change and receive no errors relevant to my code.
- [x] I have run existing tests to ensure my code doesn't break anything.
- [x] I have added tests for all relevant use cases of my code, and those tests are passing.
- [x] I have only exported functions, variables and structs that should be used from other packages.
- [x] I have added meaningful comments to all exported functions, variables, and structs.

### Website/documentation checklist

- [ ] I have locally started the website as [described here](https://github.com/opentofu/opentofu/blob/main/website/README.md) and checked my changes.

### CODE REVIEW

File: internal/command/test.go [hunk 1]
```diff
@@ -1266,3 +1268,22 @@ func (runner *TestFileRunner) prepareInputVariablesForAssertions(config *configs
 		config.Module.Variables = currentVars
 	}, diags
 }
+
+func checkPromblematicPlanErrors(run *moduletest.Run, planDiags tfdiags.Diagnostics) tfdiags.Diagnostics {
+	expectedFailures, _ := run.BuildExpectedFailuresAndSourceMaps()
```
[the remark this answers is not in the document]
Comment (Andrew-Hayes) [reply internal/command/test.go:1273]: good plan

File: internal/command/test.go [hunk 2]
```diff
@@ -1266,3 +1268,22 @@ func (runner *TestFileRunner) prepareInputVariablesForAssertions(config *configs
...
+
+	for _, diag := range planDiags {
+		if rule, ok := addrs.DiagnosticOriginatesFromCheckRule(diag); ok {
+			if rule.Container.CheckableKind() == addrs.CheckableInputVariable {
+				addr := rule.Container.(addrs.AbsInputVariableInstance)
```
[the remark this answers is not in the document]
Comment (Andrew-Hayes) [reply internal/command/test.go:1278]: i'd thought the same, but we don't in `ValidateExpectedFailures` so figured it was safe. But yeah, i'll add it in

**Review by ollevche — APPROVED** [review]
Looks good, thanks!

**Comment by ollevche** [comment]
Oh, we should also fix the related failed tests.

File: internal/command/test.go [hunk 3]
```diff
@@ -1266,3 +1271,22 @@ func (runner *TestFileRunner) prepareInputVariablesForAssertions(config *configs
...
+			if rule.Container.CheckableKind() == addrs.CheckableInputVariable {
+				addr, ok := rule.Container.(addrs.AbsInputVariableInstance)
+				if ok && expectedFailures.Has(addr.Variable) {
+					planDiags = planDiags.Append(tfdiags.Sourceless(
+						tfdiags.Warning,
+						"Invalid Variable in test file",
+						fmt.Sprintf("Variable %s, has an invalid value within the test. Although this was an expected failure, it has meant the apply stage was unable to run so the overall test will fail.", rule.Container.String())))
+				}
+			}
+		}
```
Comment (ollevche) [inline internal/command/test.go:1289]: nit: would be much more readable if we flatten it a bit:

```suggestion
		rule, ok := addrs.DiagnosticOriginatesFromCheckRule(diag)
		if !ok {
			continue
		}
		if rule.Container.CheckableKind() != addrs.CheckableInputVariable {
			continue
		}
		
		addr, ok := rule.Container.(addrs.AbsInputVariableInstance)
		if ok && expectedFailures.Has(addr.Variable) {
			planDiags = planDiags.Append(tfdiags.Sourceless(
				tfdiags.Warning,
				"Invalid Variable in test file",
				fmt.Sprintf("Variable %s, has an invalid value within the test. Although this was an expected failure, it has meant the apply stage was unable to run so the overall test will fail.", rule.Container.String())))
		}
```

**Review by ollevche — APPROVED** [review]

**Pull request merged**